### PR TITLE
Support Firecracker clean teardown - prerequisite for rust sanitizers and/or Valgrind

### PR DESCRIPTION
**Pull request by acatangiu**
# Reason for This PR

1. We want to simplify Rust integration tests to allow them to run under `cargo` instead of being forked as separate processes. This drastically improves maintenance and debuggability (with old `fork()` approach we have no visibility into why a test fails, we just see that it does).
2. We want to be able to run [Rust sanitizers](https://rustc-dev-guide.rust-lang.org/sanitizers.html) on the Firecracker unit and integration tests for further code-base/security hardening.
3. #119 we might want to add `valgrind` test to check for memory leaks around our FFI interfaces.

## Description of Changes

This PR is based on #2535 and some of the commits are co-authored by `Cerul Alain <ae@metaeducation.com>`. 

This PR adds multiple incremental commits all working toward one goal: to allow Firecracker to propagate any exit reason all the way to the top level `main()` function and have a single program exit point while running all destructors and cleanup code during teardown of all resources.

For this to be possible, the following high-level steps were taken (per-step details enclosed in their respective commit):
- slightly reworked `Vcpu` teardown to allow thread main function to end instead of blocking forever,
- added internal api-endpoint used to trigger API thread teardown,
- reworked the Vmm stopping/exit logic to remove existing cyclic dependencies that would not allow destructors to be run,
- `Vmm` manages and exposes an `shutdown_exit_code` that is used by the higher layers to stop the main loop in when `Vmm` is shut down,
- many `libc::exit()` points across the codebase have been unified into a natural flow of propagating errors to above layers with `main()` being responsible for exiting the process,
- seccomp filters adjusted to allow syscalls done by Rust during process cleanup.

End result is:
- a cleaner model that's easier to understand and work with,
- cleaner Rust integration tests that are now easy to debug when failing,
- removed weird workarounds like ApiServer running special logic to interpret Vmm responses to decide if it needs to kill process after responding
- clean, cycle-free `Vcpu` and `Vmm` stopping protocol

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] ~Any required documentation changes (code and docs) are included in this PR.~
- [x] ~Any newly added `unsafe` code is properly documented.~
- [x] ~Any API changes are reflected in `firecracker/swagger.yaml`.~
- [x] ~Any user-facing changes are mentioned in `CHANGELOG.md`.~
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: tests/framework/microvm.py [hunk 1]
```diff
@@ -144,18 +144,31 @@ def __init__(
         # External clone/exec tool, because Python can't into clone
         self.bin_cloner_path = bin_cloner_path
 
+        # Flag checked in destructor to see abnormal signal-induced crashes.
+        self.expect_kill_by_signal = False
```
Comment (gc-plp) [inline tests/framework/microvm.py:148]: Wouldn't it make sense to have this always on?

File: tests/framework/microvm.py [hunk 2]
```diff
@@ -144,18 +144,31 @@ def __init__(
...
+
     def kill(self):
         """All clean up associated with this microVM should go here."""
         # pylint: disable=subprocess-run-check
         if self.logging_thread is not None:
             self.logging_thread.stop()
 
+        if self.expect_kill_by_signal is False and \
+                "Shutting down VM after intercepting signal" in self.log_data:
+            # Too late to assert at this point, pytest will still report the
```
Comment (gc-plp) [inline tests/framework/microvm.py:158]: I think that before the test is marked as passed/failed, we can add a check in `scheduler.py` that takes a look at what microvms are running and check for problems. This way we can forcibly fail the test if something unexpected happens and the test is marked as passed.

File: tests/framework/microvm.py (hunk 1 above)
Comment (acatangiu) [reply tests/framework/microvm.py:148]: No.. default in tests and here is clean exit and not being killed by a signal.

Some tests validate/expect the microvm to be killed by a signal, this flag is set to `True` by those tests so microvm destructor knows to not consider it an error.

File: tests/framework/microvm.py [hunk 3]
```diff
@@ -144,18 +144,31 @@ def __init__(
...
+            # test as passed. BUT we can dump full logs for debugging,
+            # as well as an intentional eye-sore in the test report.
+            print(self.log_data)
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply tests/framework/microvm.py:161]: right, will fix!

File: tests/framework/microvm.py (hunk 2 above)
Comment (acatangiu) [reply tests/framework/microvm.py:158]: > What I'm proposing is slightly more complex, so we can go down the path from this PR and track the improvement as an issue. WDYT?

++

**Comment by berciuliviu** [comment]
Really great job with this PR! Things are clear and well explained. I have tried running the test `test_failing_filter` on my machine, but it passes everytime. Maybe a rebase with the latest changes on main and a test-rerun could do the trick?

File: src/vmm/src/lib.rs [hunk 4]
```diff
@@ -355,25 +363,12 @@ impl Vmm {
...
-        if let Err(e) = METRICS.write() {
-            error!("Failed to write metrics while stopping: {}", e);
-        }
-
-        // Exit from Firecracker using the provided exit code. Safe because we're terminating
-        // the process anyway.
-        unsafe {
-            libc::_exit(exit_code);
+        self.shutdown_exit_code = Some(exit_code);
+        if let Err(e) = self.exit_evt.write(1) {
```
Comment (alindima) [inline src/vmm/src/lib.rs]: If I understand correctly, this function is called when the `exit_evt` was signaled. And in this same function, we signal `exit_evt` again and set the exit code, which seems a little redundant to me. Can't we do all of this in the Subscriber::process implementation?

Is this a workaround to the fact that there is currently no way of breaking out of the event manager loop?

**Review by alindima — COMMENTED** [review]
I still got a couple more commits to review, but looking good so far.

I am wondering whether there is significant impact on the shutdown time. Do you view this as critical/something to consider?

File: src/vmm/src/lib.rs (hunk 4 above)
Comment (acatangiu) [reply src/vmm/src/lib.rs]: This is part of the mess/cyclic code that was ultimately refactored. The `Vcpus` commit will clean up and document the teardown protocol with its communication and steps. Please take a look at the final code.

> I am wondering whether there is significant impact on the shutdown time. Do you view this as critical/something to consider?

I may be experiencing a failure of imagination, but I can't think of any production scenario where process closure time would be so critical/latency-sensitive as to be impacted by graceful teardown.

**Review by berciuliviu — APPROVED** [review]
LGTM, just don't forget to replace the LOG.error() gc-plp mentioned

File: src/vmm/src/lib.rs (hunk 4 above)
Comment (alindima) [reply src/vmm/src/lib.rs]: > This is part of the mess/cyclic code that was ultimately refactored. The Vcpus commit in this PR cleans up and documents the teardown protocol with its communication and steps. Please take a look at the final code.

Cool, I usually review code commit-by-commit :D

File: src/vmm/src/lib.rs [hunk 5]
```diff
@@ -743,21 +769,29 @@ impl Subscriber for Vmm {
...
+
+            let mut exit_code = None;
+            // Query each vcpu for their exit_code.
+            for handle in &self.vcpus_handles {
+                match handle.response_receiver().try_recv() {
+                    Ok(VcpuResponse::Exited(status)) => {
+                        if exit_code.is_none() {
+                            exit_code = Some(status);
+                        } else {
+                            warn!("Multiple exit states detected, using first exit code");
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply src/vmm/src/lib.rs:784]: sure, will do

File: src/vmm/tests/integration_tests.rs [hunk 6]
```diff
@@ -59,262 +57,184 @@ fn test_build_microvm() {
...
-        }
-        vmm_pid => {
-            // Parent process: wait for the vmm to exit.
-            wait_vmm_child_process(vmm_pid);
-        }
-    }
+    let (vmm, mut _evmgr) = default_vmm(None);
+
+    // On x86_64, the vmm should exit once its workload completes and signals the exit event.
+    // On , the test kernel doesn't exit, so the vmm is force-stopped.
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply src/vmm/tests/integration_tests.rs:63]: right, will fix

File: resources/seccomp/aarch64-unknown-linux-musl.json [hunk 7]
```diff
@@ -381,6 +398,14 @@
                 "syscall": "recvfrom",
                 "comment": "Used to retrieve data from the socket"
             },
+            {
+                "syscall": "rt_sigprocmask",
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply resources/seccomp/aarch64-unknown-linux-musl.json:402]: apparently not. I believe it is only called for _child_ thread cleanup and not called during _main_ thread cleanup during process exit.

File: resources/seccomp/aarch64-unknown-linux-musl.json [hunk 8]
```diff
@@ -513,10 +551,31 @@
...
             {
                 "syscall": "rt_sigreturn",
                 "comment": "rt_sigreturn is needed in case a fault does occur, so that the signal handler can return. Otherwise we get stuck in a fault loop."
             },
+            {
+                "syscall": "sigaltstack",
+                "comment": "sigaltstack is used by Rust stdlib to remove alternative signal stack during thread teardown."
+            },
+            {
+                "syscall": "futex",
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply resources/seccomp/aarch64-unknown-linux-musl.json:567]: It is used on all threads, thread synchronization techniques like `barrier` or `join` or `channel`s use underlying `futex`

File: src/vmm/src/lib.rs (hunk 4 above)
Comment (alindima) [reply src/vmm/src/lib.rs]: > I may be experiencing a failure of imagination, but I can't think of any production scenario where process closure time would be so critical/latency-sensitive as to be impacted by graceful teardown.

I think you're right. Anyone who wants to quickly kill the Firecracker process can just send a SIGKILL.

File: tests/framework/microvm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (acatangiu) [reply tests/framework/microvm.py:161]: @gc-plp I tried using `LOG.error()` but it doesn't report anything for `PASSED` tests, thus hiding any silent seccomp triggers..

The `print()` always ends up in the test report however.

File: resources/seccomp/x86_64-unknown-linux-musl.json [hunk 9]
```diff
@@ -70,6 +70,14 @@
                 "syscall": "rt_sigreturn",
                 "comment": "rt_sigreturn is needed in case a fault does occur, so that the signal handler can return. Otherwise we get stuck in a fault loop."
             },
+            {
+                "syscall": "sigaltstack",
+                "comment": "sigaltstack is used by Rust stdlib to remove alternative signal stack during thread teardown."
+            },
+            {
+                "syscall": "tkill",
```
Comment (alindima) [inline resources/seccomp/x86_64-unknown-linux-musl.json:78]: Can we tighten this with an argument check?

Comment (acatangiu) [reply resources/seccomp/x86_64-unknown-linux-musl.json:78]: yes, good callout!

File: tests/framework/microvm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gc-plp) [reply tests/framework/microvm.py:161]: I think that we can activate the live logging feature and raise the default logging level to ERROR:

```diff
diff --git a/tests/pytest.ini b/tests/pytest.ini
index e980044a..d1070b0d 100644
--- a/tests/pytest.ini
+++ b/tests/pytest.ini
@@ -13,7 +13,7 @@ timeout = 300
 cache_dir = ../build/pytest_cache
 ; Set logger format and level
-log_level = INFO
+log_level = ERROR
 log_format = %(asctime)s %(name)s: %(levelname)s %(message)s
 ; Uncomment the line below to get the live logger output
-;log_cli = true
+log_cli = true
```

[the remark this answers is not in the document]
Comment (gc-plp) [reply tests/framework/microvm.py:161]: Actually here's a better version that keeps everything in the background while printing errors in the console:

```diff
diff --git a/tests/pytest.ini b/tests/pytest.ini
index e980044a..af57aab3 100644
--- a/tests/pytest.ini
+++ b/tests/pytest.ini
@@ -15,5 +15,6 @@ cache_dir = ../build/pytest_cache
 ; Set logger format and level
 log_level = INFO
 log_format = %(asctime)s %(name)s: %(levelname)s %(message)s
-; Uncomment the line below to get the live logger output
-;log_cli = true
+
+log_cli_level = ERROR
+log_cli = true

```

**Comment by acatangiu** [comment]
@berciuliviu @alindima addressed all comments, PTAL

**Review by alindima — APPROVED** [review]
Great job! LGTM

**Review by berciuliviu — APPROVED** [review]
LGTM!

**Review by berciuliviu — APPROVED** [review]

**Review by alindima — APPROVED** [review]

**Comment by AE1020** [comment]
Glad it worked out!

**Pull request merged**